### DISCUSSION TITLE
Custom fields exposed by device (does not work in HA)

### DISCUSSION THREAD

**Discussion by grafalex82** [Koenkk/zigbee2mqtt#9285] [Devices]
Hi All,

I am working on a device, that implements a custom cluster (some extension on genOnOffSwitchCfg),  providing several enum settings to the user. I would like them to be editable/selectable via home assistant (everything is fine in z2m dashboard).

In the exposes section I am defining settings pretty much in the same way ([full code here](https://github.com/grafalex82/hellozigbee/blob/hello_zigbee_custom_cluster/myswitch.js#L154) )

```
const switchTypeValues = ['toggle', 'momentary', 'multifunction'];
const switchActionValues = ['onOff', 'offOn', 'toggle'];
...
[
        exposes.enum('switch_type', ea.ALL, switchTypeValues).withEndpoint(epName),
        exposes.enum('switch_actions', ea.ALL, switchActionValues).withEndpoint(epName),
]

```

In Z2M dashboard both fields show normally
![image](https://user-images.githubusercontent.com/23584882/138516071-9106a4ac-52ce-4bf8-bbe2-d53834c40956.png)

When device is being interviewed and registered, Z2M sends multiple MQTT events describing device itself, and its fields in particular. I have not found any fundamental differences between switch_type and switch_actions descriptions. Unfortunately HA displays just switch_actions field, while hide/disable switch_type.

![image](https://user-images.githubusercontent.com/23584882/138517847-98d7dcf0-0b14-4b94-be3f-e1dddb5ace7b.png)

HA logs does not show anything suspicious.

What could be wrong with switch_type definition so that it is not visible/editable in home assistant?

**Comment by Koenkk** [comment]
It looks the entity is disabled by default, you can enable it by clicking the toggle on your screenshot.

**Reply by grafalex82** [reply]
Thank you, Koen!
Indeed. I enabled those entities. Some enabled fine and got a combobox for values selection. Unfortunately other items display as non-editable label. Interestingly, fields of the same type for some reason get different icons.

Do you by any chance know how HA decide to enable/disable entities, and how it selects icons? Is there a chance to control this from the converter?

**Reply by grafalex82** [reply]
I found [this piece of code](https://github.com/Koenkk/zigbee2mqtt/blob/master/lib/extension/homeassistant.ts#L466-L508), that disables most of the enum entities by default. Is there a way to control enabled_by_default and icon fields from the converter somehow?

**Reply by Koenkk** [reply]
You can override the Home Assistant discovery payload via the `configuration.yaml`.

Example: https://www.zigbee2mqtt.io/integration/home_assistant.html#exposing-switch-as-a-light

**Reply by grafalex82** [reply]
Koen, thank you for the link. I'll try.
